lib/epochs.js

```diff
@@ -78,27 +78,72 @@ module.exports = function Epochs(ssb) {
     },
   }
 
-  return {
-    getEpochs(groupId, cb) {
-      if (cb === undefined) return p(this.getEpochs)(groupId)
+  function getEpochs(groupId, cb) {
+    if (cb === undefined) return p(getEpochs)(groupId)
+
+    const getters = pluck(allGetters, ['author', 'secret'])
+    reduceEpochs(ssb, groupId, getters, (err, reduce) => {
+      if (err) return cb(clarify(err, 'Failed to resolve epoch @tangle/reduce'))
+
+      const epochs = reduce.graph.connectedNodes.map((node) => {
+        const info = {
+          id: node.key, // alias: epochRootId
+          previous: node.previous,
+          ...node.data[node.key],
+        }
+        info.secret = Buffer.from(info.secret, 'base64')
+        return info
+      })
 
-      const getters = pluck(allGetters, ['author', 'secret'])
-      reduceEpochs(ssb, groupId, getters, (err, reduce) => {
-        if (err)
-          return cb(clarify(err, 'Failed to resolve epoch @tangle/reduce'))
+      cb(null, epochs)
+    })
+  }
 
-        const epochs = reduce.graph.connectedNodes.map((node) => {
-          const info = {
-            id: node.key, // alias: epochRootId
-            previous: node.previous,
-            ...node.data[node.key],
-          }
-          info.secret = Buffer.from(info.secret, 'base64')
-          return info
+  return {
+    getEpochs,
+
+    /**
+     * Get the epoch matching the currently picked write key for the group.
+     */
+    getPickedEpoch(groupId, opts) {
+      let lastKey
+
+      return pull(
+        opts?.live
+          ? ssb.box2.getGroupInfoUpdates(groupId)
+          : pull(ssb.box2.getGroupInfoUpdates(groupId), pull.take(1)),
+        pull.map((groupInfo) => {
+          // prettier-ignore
+          if (groupInfo.excluded) throw new Error("Can't get picked epoch for group we're excluded from")
+
+          return groupInfo
+        }),
+        // only pass forward group info if the writeKey is different to last time
+        pull.filter((groupInfo) => {
+          if (lastKey?.equals(groupInfo.writeKey.key)) return false
+
+          lastKey = groupInfo.writeKey.key
+          return true
+        }),
+        // find details about the newly picked epoch
+        pull.asyncMap((groupInfo, cb) => {
+          getEpochs(groupId, (err, epochs) => {
+            // prettier-ignore
+            if (err) return cb(clarify(err, "Error getting info about epochs when looking for picked epoch"))
+
```

Comment:
was getting errors so added a log on this line
```
            console.log('got epochs but no err', epochs, 'groupInfo', groupInfo)
```

and got these logs. it's flaky though so happens on like 50% of runs

```
...
got epochs but no err [
  {
    id: 'ssb:message/classic/HQJyiWkVCVosHwKQPmfYkcShF0jWZOBnozNQQ6fPTc0=',
    previous: null,
    author: 'ssb:feed/bendybutt-v1/R7oVP6uSBqSmyC-iQxX67S9Zl2498yxJk4NarqMaBtI=',
    secret: <Buffer 39 59 6c 7d fc 27 3d d7 49 42 ec ec 13 77 5b ae 5d b4 93 17 9c 48 c1 05 8a 40 01 79 2c a2 dc 8c>
  }
] groupInfo {
  readKeys: [
    {
      key: <Buffer 39 59 6c 7d fc 27 3d d7 49 42 ec ec 13 77 5b ae 5d b4 93 17 9c 48 c1 05 8a 40 01 79 2c a2 dc 8c>,
      scheme: 'envelope-large-symmetric-group'
    },
    {
      key: <Buffer 6b bd f6 bb 5f f3 61 58 dd 46 3d 65 63 e1 4d e0 3e 68 76 77 6a 4d 95 a6 35 7b 01 f4 4e 83 c7 10>,
      scheme: 'envelope-large-symmetric-group'
    }
  ],
  writeKey: {
    key: <Buffer 6b bd f6 bb 5f f3 61 58 dd 46 3d 65 63 e1 4d e0 3e 68 76 77 6a 4d 95 a6 35 7b 01 f4 4e 83 c7 10>,
    scheme: 'envelope-large-symmetric-group'
  },
  root: 'ssb:message/classic/HQJyiWkVCVosHwKQPmfYkcShF0jWZOBnozNQQ6fPTc0='
}
pickedEPoch undefined
not ok 5 TypeError: Cannot read properties of undefined (reading 'id')
  ---
    operator: error
    at: <anonymous> (/home/me/prj/ssb/ssb-tribes2/test/list-members.test.js:204:18)
    stack: |-
      TypeError: Cannot read properties of undefined (reading 'id')
          at /home/me/prj/ssb/ssb-tribes2/index.js:326:47
          at /home/me/prj/ssb/ssb-tribes2/node_modules/.pnpm/pull-stream@3.7.0/node_modules/pull-stream/throughs/map.js:13:23
          at /home/me/prj/ssb/ssb-tribes2/node_modules/.pnpm/pull-stream@3.7.0/node_modules/pull-stream/throughs/async-map.js:39:20
          at /home/me/prj/ssb/ssb-tribes2/lib/epochs.js:137:13
          at /home/me/prj/ssb/ssb-tribes2/lib/epochs.js:98:7
          at /home/me/prj/ssb/ssb-tribes2/lib/epochs.js:333:13
          at /home/me/prj/ssb/ssb-tribes2/node_modules/.pnpm/pull-stream@3.7.0/node_modules/pull-stream/sinks/reduce.js:10:5
          at /home/me/prj/ssb/ssb-tribes2/node_modules/.pnpm/pull-stream@3.7.0/node_modules/pull-stream/sinks/drain.js:26:24
          at /home/me/prj/ssb/ssb-tribes2/node_modules/.pnpm/pull-stream@3.7.0/node_modules/pull-stream/throughs/async-map.js:28:19
          at /home/me/prj/ssb/ssb-tribes2/node_modules/.pnpm/pull-stream@3.7.0/node_modules/pull-stream/sources/values.js:19:7
  ...
```

it's weird that getEpochs returns fewer epochs/secrets than getGroupInfoUpdates. a bug in getEpochs? @mixmix